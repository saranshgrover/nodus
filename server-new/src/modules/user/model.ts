import { getModelForClass } from '@typegoose/typegoose'
import { ObjectId } from 'mongodb'

import { User } from '../../entities'

// This generates the mongoose model for us
export const UserMongooseModel = getModelForClass(User)

export default class UserModel {
	async getById(_id: ObjectId): Promise<User | null> {
		// Use mongoose as usual
		return UserMongooseModel.findById(_id).lean().exec()
	}
}
