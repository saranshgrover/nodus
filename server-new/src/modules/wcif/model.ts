import { DocumentType, getModelForClass } from '@typegoose/typegoose'
import { ObjectId } from 'mongodb'
import { Wcif } from '../../entities/wcif'

// This generates the mongoose model for us
export const WcifMongooseModel = getModelForClass(Wcif)

export default class WcifModel {
	async getById(_id: ObjectId): Promise<Wcif | null> {
		// Use mongoose as usual
		return WcifMongooseModel.findById(_id).lean().exec()
	}
	async create(data: Wcif): Promise<DocumentType<Wcif>> {
		const wcif = new WcifMongooseModel(data)
		await wcif.save()
		return wcif
	}

	async findByCompetitionId(
		competitionId: string
	): Promise<DocumentType<Wcif> | null> {
		return await WcifMongooseModel.findOne({ competitionId })
	}

	async findByCompetitionIdLean(competitionId: string) {
		return await WcifMongooseModel.findOne({ competitionId }).lean().exec()
	}

	async exists(competitionId: string): Promise<boolean> {
		return await WcifMongooseModel.exists({ competitionId: competitionId })
	}

	async findAll(): Promise<Wcif[]> {
		return await WcifMongooseModel.find().lean().exec()
	}

	async deleteWcif(competitionId: string): Promise<Wcif | null> {
		return await WcifMongooseModel.findOneAndRemove({
			competitionId,
		}).exec()
	}
}
