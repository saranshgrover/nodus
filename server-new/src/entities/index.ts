// Don't forget to export your entities here to have cleaner imports
export * from './user'
export * from './wcif'
