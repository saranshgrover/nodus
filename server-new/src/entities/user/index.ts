export { Competition } from './competition'
export { ExternalConnection } from './externalConnection'
export { Notification } from './notification'
export { User } from './user'
export { WCAContent, WCATeams } from './WCAContent'
