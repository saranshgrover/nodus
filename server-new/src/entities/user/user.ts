import { ObjectType, Field } from 'type-graphql'
import { prop } from '@typegoose/typegoose'
import { ObjectId } from 'mongodb'

import { Competition } from './competition'
import { ExternalConnection } from './externalConnection'

@ObjectType()
export class User {
	@Field()
	readonly _id!: ObjectId

	@prop()
	@Field()
	username!: string

	@prop()
	hashedPassword: string

	@prop()
	@Field()
	primaryAuthenticationType: string

	@prop()
	@Field(() => [ExternalConnection])
	competitions: Competition[]
}
