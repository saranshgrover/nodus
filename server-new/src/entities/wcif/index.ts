export { Activity } from './activity'
export { AdvancementCondition } from './advancementCondition'
export { Assignment } from './assignment'
export { Attempt } from './attempt'
export { Avatar } from './avatar'
export { ChildActivity } from './childActivity'
export { Cutoff } from './cutoff'
export { Event } from './event'
export { Extension } from './extension'
export { Person } from './person'
export { PersonalBest } from './personalBest'
export { Qualification } from './qualification'
export { Registration } from './registration'
export { Result } from './result'
export { Room } from './room'
export { Round } from './round'
export { Schedule } from './schedule'
export { ScrambleSet } from './scrambleSet'
export { Setting } from './setting'
export { TimeLimit } from './timeLimit'
export { Venue } from './venue'
export { Wcif } from './wcif'
export { WcifFetch } from './wcifFetch'
