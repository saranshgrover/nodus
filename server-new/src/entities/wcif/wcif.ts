import { ObjectType, InputType, Field, Int } from 'type-graphql'
import { prop } from '@typegoose/typegoose'
import { ObjectId } from 'mongodb'
import { Extension } from './extension'
import { Person } from './person'
import { Event } from './event'
import { Schedule } from './schedule'
import { Setting } from './setting'

@InputType('WcifInput')
@ObjectType('Wcif')
export class Wcif {
	@Field()
	readonly _id!: ObjectId

	@prop()
	@Field(() => String)
	competitionid: string

	@prop()
	@Field()
	name!: string

	@prop()
	@Field()
	shortName: string

	@prop()
	@Field(() => [Person])
	persons: Person[]

	@prop()
	@Field(() => [Event])
	events: Event[]

	@prop()
	@Field(() => Schedule)
	schedule: Schedule

	@prop()
	@Field(() => Int)
	competitorLimit: number

	@prop()
	@Field(() => [Extension])
	extensions: Extension[]

	@prop()
	@Field()
	locationName!: string

	@prop()
	@Field()
	registrationOpen!: string

	@prop()
	@Field()
	registrationClose!: string

	@prop()
	@Field(() => [Setting])
	settings!: Setting[]
}
