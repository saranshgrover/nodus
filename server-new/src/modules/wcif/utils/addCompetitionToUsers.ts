import { Wcif } from '../../../entities'
import { UserMongooseModel } from '../../user/model'

export default async function addCompetitionToUsers(wcif: Wcif) {
	const persons = wcif.persons
	for (const person of persons) {
		const user = await UserMongooseModel.findOne({
			connections: {
				$elemMatch: {
					$and: [{ connectionType: 'WCA' }, { 'content.id': person.wcaUserId }],
				},
			},
		})
		if (
			user &&
			user.competitions &&
			!user.competitions.some(
				(competition) => competition.competitionId === wcif.competitionId
			)
		) {
			let endDate = new Date(wcif.schedule.startDate)
			endDate.setDate(endDate.getDate() + wcif.schedule.numberOfDays - 1)
			const roles: RoleType[] = ['competitor']
			if (person.roles.includes('organizer')) roles.push('organizer')
			if (person.roles.includes('delegate')) roles.push('delegate')
			if (person.roles.includes('delegate')) roles.push('delegate')
			if (person.roles.includes('trainee_delegate'))
				roles.push('traineeDelegate')
			user.competitions.push({
				competitionType: 'WCA',
				competitionId: wcif.competitionId,
				startDate: wcif.schedule.startDate,
				endDate: endDate.toISOString().split('T')[0],
				roles: roles,
			})
			await user.save()
		}
	}
}
