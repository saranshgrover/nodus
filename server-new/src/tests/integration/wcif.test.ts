import { ApolloServer, gql } from 'apollo-server-express'
import { createTestClient } from 'apollo-server-testing'
import { Wcif } from '../../entities'
import { UserMongooseModel } from '../../modules/user/model'
import { WcifMongooseModel } from '../../modules/wcif/model'
import addCompetitionToUsers from '../../modules/wcif/utils/addCompetitionToUsers'
import createModelFromWcif from '../../modules/wcif/utils/createModelFromWcif'
import { buildSchema } from '../../utils'
import { createWcaUser } from '../data/user-builder'
import { Competition, Person } from '../data/wcif-builders'
import wcif1 from '../data/wcif1'
import { closeDatabase, connect, populateDatabase } from '../utils'

beforeAll(async () => connect())

beforeEach(async () => {
	await populateDatabase(WcifMongooseModel, [wcif1])
})

// afterEach(async () => {
// 	await clearDatabase()
// })

afterAll(async (done) => {
	await closeDatabase()
	done()
})

/**
 * Prompt test suite.
 */
describe('Wcif', () => {
	it(`should throw not logged in error`, async () => {
		// We build the schema
		const graphQLSchema = await buildSchema()

		// We create the Apollo Server
		const server = new ApolloServer({
			schema: graphQLSchema,
			context: () => ({ req: {} }),
		}) as any

		// use the test server to create a query function
		const { mutate } = createTestClient(server)

		// We define the query and the variables as you would do from your front-end
		const variables = {
			competitionId: 'SBUFall2019',
		}

		const CREATE_WCIF = gql`
			mutation createTodo($competitionId: String!) {
				createWcif(competitionId: $competitionId) {
					competitionId
					_id
				}
			}
		`

		// run query against the server and snapshot the output
		const res = await mutate({
			mutation: CREATE_WCIF,
			variables,
		})

		expect(res).toMatchSnapshot()
	})

	it(`should  create a wcif`, async () => {
		//@ts-ignore
		const wcif = await createModelFromWcif(wcif1, 'SBUFall2019')
		// this value is dynamic, so snapshot wont match it
		// @ts-ignore
		delete wcif.synchronizedAt
		expect(wcif).toMatchSnapshot()
	})

	it(`should add competition to users competition list`, async () => {
		const user = createWcaUser()
		const competition: Wcif = Competition({
			persons: [Person({ wcaUserId: user.connections[0].content.id })],
		})
		await populateDatabase(UserMongooseModel, [user])
		await addCompetitionToUsers(competition)
		await populateDatabase(WcifMongooseModel, [competition])

		// We create the Apollo Server
		const graphQLSchema = await buildSchema()
		const server = new ApolloServer({
			schema: graphQLSchema,
			context: () => ({ req: {} }),
		}) as any

		// use the test server to create a query function
		const { query } = createTestClient(server)

		const GET_USER = gql`
			query getUser($id: ObjectId!) {
				getUser(id: $id) {
					competitions {
						competitionId
						roles
					}
				}
			}
		`
		const res = await query({
			query: GET_USER,
			variables: { id: user._id.toHexString() },
		})
		expect(res.data?.getUser.competitions).toEqual([
			{
				competitionId: competition.competitionId,
				roles: ['competitor'],
			},
		])
	})
})
