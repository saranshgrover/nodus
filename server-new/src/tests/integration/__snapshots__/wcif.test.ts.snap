// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Wcif should  create a wcif 1`] = `
Object {
  "competitionId": "SBUFall2019",
  "competitorLimit": 250,
  "events": Array [
    Object {
      "extensions": Array [],
      "id": "333",
      "rounds": Array [
        Object {
          "advancementCondition": Object {
            "level": 64,
            "type": "ranking",
          },
          "cutoff": null,
          "extensions": Array [],
          "format": "a",
          "id": "333-r1",
          "results": Array [
            Object {
              "attempts": Array [
                Object {
                  "reconstruction": null,
                  "result": 623,
                },
                Object {
                  "reconstruction": null,
                  "result": 1047,
                },
                Object {
                  "reconstruction": null,
                  "result": 777,
                },
                Object {
                  "reconstruction": null,
                  "result": 708,
                },
                Object {
                  "reconstruction": null,
                  "result": 909,
                },
              ],
              "average": 798,
              "best": 623,
              "personId": 134,
              "ranking": 1,
            },
            Object {
              "attempts": Array [
                Object {
                  "reconstruction": null,
                  "result": 730,
                },
                Object {
                  "reconstruction": null,
                  "result": 983,
                },
                Object {
                  "reconstruction": null,
                  "result": 746,
                },
                Object {
                  "reconstruction": null,
                  "result": 942,
                },
                Object {
                  "reconstruction": null,
                  "result": 731,
                },
              ],
              "average": 806,
              "best": 730,
              "personId": 29,
              "ranking": 2,
            },
          ],
          "scrambleSetCount": 6,
          "timeLimit": Object {
            "centiseconds": 60000,
            "cumulativeRoundIds": Array [],
          },
        },
      ],
    },
  ],
  "extensions": Array [],
  "formatVersion": "1.0",
  "locationName": "Student Activities Center, Stony Brook, NY 11790, USA",
  "name": "SBU Fall 2019",
  "persons": Array [
    Object {
      "assignments": Array [
        Object {
          "activityId": 71,
          "assignmentCode": "competitor",
          "stationNumber": null,
        },
        Object {
          "activityId": 63,
          "assignmentCode": "competitor",
          "stationNumber": null,
        },
        Object {
          "activityId": 75,
          "assignmentCode": "competitor",
          "stationNumber": null,
        },
        Object {
          "activityId": 85,
          "assignmentCode": "competitor",
          "stationNumber": null,
        },
        Object {
          "activityId": 57,
          "assignmentCode": "competitor",
          "stationNumber": null,
        },
        Object {
          "activityId": 38,
          "assignmentCode": "competitor",
          "stationNumber": null,
        },
        Object {
          "activityId": 65,
          "assignmentCode": "staff-scrambler",
          "stationNumber": null,
        },
        Object {
          "activityId": 45,
          "assignmentCode": "competitor",
          "stationNumber": null,
        },
      ],
      "avatar": Object {
        "thumbUrl": "https://staging.worldcubeassociation.org/uploads/user/avatar/2015LARS02/1574949155_thumb.jpeg",
        "url": "https://staging.worldcubeassociation.org/uploads/user/avatar/2015LARS02/1574949155.jpeg",
      },
      "birthdate": "1954-12-04",
      "countryIso2": "US",
      "email": "126272@worldcubeassociation.org",
      "gender": "m",
      "name": "Harry Larson",
      "personalBests": Array [
        Object {
          "best": 504,
          "continentalRanking": 3372,
          "eventId": "222",
          "nationalRanking": 2429,
          "type": "average",
          "worldRanking": 12318,
        },
        Object {
          "best": 1349,
          "continentalRanking": 2665,
          "eventId": "333",
          "nationalRanking": 1950,
          "type": "average",
          "worldRanking": 11271,
        },
        Object {
          "best": 4746,
          "continentalRanking": 5423,
          "eventId": "333oh",
          "nationalRanking": 3664,
          "type": "average",
          "worldRanking": 23205,
        },
        Object {
          "best": 6562,
          "continentalRanking": 2982,
          "eventId": "444",
          "nationalRanking": 2060,
          "type": "average",
          "worldRanking": 12858,
        },
        Object {
          "best": 1292,
          "continentalRanking": 6459,
          "eventId": "pyram",
          "nationalRanking": 4760,
          "type": "average",
          "worldRanking": 24886,
        },
        Object {
          "best": 368,
          "continentalRanking": 4640,
          "eventId": "222",
          "nationalRanking": 3284,
          "type": "single",
          "worldRanking": 17465,
        },
        Object {
          "best": 1065,
          "continentalRanking": 2234,
          "eventId": "333",
          "nationalRanking": 1639,
          "type": "single",
          "worldRanking": 9137,
        },
        Object {
          "best": 3531,
          "continentalRanking": 4962,
          "eventId": "333oh",
          "nationalRanking": 3345,
          "type": "single",
          "worldRanking": 21828,
        },
        Object {
          "best": 5962,
          "continentalRanking": 3257,
          "eventId": "444",
          "nationalRanking": 2246,
          "type": "single",
          "worldRanking": 14039,
        },
        Object {
          "best": 16550,
          "continentalRanking": 3729,
          "eventId": "555",
          "nationalRanking": 2641,
          "type": "single",
          "worldRanking": 14591,
        },
        Object {
          "best": 969,
          "continentalRanking": 7468,
          "eventId": "pyram",
          "nationalRanking": 5485,
          "type": "single",
          "worldRanking": 28546,
        },
      ],
      "registrantId": 1,
      "registration": Object {
        "comments": "",
        "eventIds": Array [
          "222",
          "333",
          "333oh",
          "444",
          "555",
          "pyram",
        ],
        "guests": 0,
        "status": "accepted",
        "wcaRegistrationId": 363859,
      },
      "roles": Array [],
      "wcaId": "2015LARS02",
      "wcaUserId": 126272,
    },
  ],
  "registrationClose": "2019-11-16T00:00:00.000Z",
  "registrationOpen": "2019-09-21T00:00:00.000Z",
  "schedule": Object {
    "numberOfDays": 1,
    "startDate": "2019-11-23",
    "venues": Array [
      Object {
        "countryIso2": "US",
        "id": 1,
        "latitudeMicrodegrees": 40914506,
        "longitudeMicrodegrees": -73123564,
        "name": "Stony Brook University",
        "rooms": Array [
          Object {
            "activities": Array [
              Object {
                "activityCode": "pyram-r1",
                "childActivities": Array [
                  Object {
                    "activityCode": "pyram-r1-g1",
                    "childActivities": Array [],
                    "endTime": "2019-11-23T14:45:00Z",
                    "extensions": Array [],
                    "id": 81,
                    "name": "Pyraminx, Round 1, Group 1",
                    "startTime": "2019-11-23T14:30:00Z",
                  },
                  Object {
                    "activityCode": "pyram-r1-g3",
                    "childActivities": Array [],
                    "endTime": "2019-11-23T15:00:00Z",
                    "extensions": Array [],
                    "id": 83,
                    "name": "Pyraminx, Round 1, Group 3",
                    "startTime": "2019-11-23T14:45:00Z",
                  },
                  Object {
                    "activityCode": "pyram-r1-g5",
                    "childActivities": Array [],
                    "endTime": "2019-11-23T15:15:00Z",
                    "extensions": Array [],
                    "id": 85,
                    "name": "Pyraminx, Round 1, Group 5",
                    "startTime": "2019-11-23T15:00:00Z",
                  },
                  Object {
                    "activityCode": "pyram-r1-g7",
                    "childActivities": Array [],
                    "endTime": "2019-11-23T15:30:00Z",
                    "extensions": Array [],
                    "id": 87,
                    "name": "Pyraminx, Round 1, Group 7",
                    "startTime": "2019-11-23T15:15:00Z",
                  },
                ],
                "endTime": "2019-11-23T15:30:00Z",
                "extensions": Array [],
                "id": 1,
                "name": "Pyraminx, Round 1",
                "startTime": "2019-11-23T14:30:00Z",
              },
              Object {
                "activityCode": "222-r1",
                "childActivities": Array [
                  Object {
                    "activityCode": "222-r1-g1",
                    "childActivities": Array [],
                    "endTime": "2019-11-23T15:42:00Z",
                    "extensions": Array [],
                    "id": 51,
                    "name": "2x2x2 Cube, Round 1, Group 1",
                    "startTime": "2019-11-23T15:30:00Z",
                  },
                  Object {
                    "activityCode": "222-r1-g3",
                    "childActivities": Array [],
                    "endTime": "2019-11-23T15:54:00Z",
                    "extensions": Array [],
                    "id": 53,
                    "name": "2x2x2 Cube, Round 1, Group 3",
                    "startTime": "2019-11-23T15:42:00Z",
                  },
                  Object {
                    "activityCode": "222-r1-g5",
                    "childActivities": Array [],
                    "endTime": "2019-11-23T16:06:00Z",
                    "extensions": Array [],
                    "id": 55,
                    "name": "2x2x2 Cube, Round 1, Group 5",
                    "startTime": "2019-11-23T15:54:00Z",
                  },
                  Object {
                    "activityCode": "222-r1-g7",
                    "childActivities": Array [],
                    "endTime": "2019-11-23T16:18:00Z",
                    "extensions": Array [],
                    "id": 57,
                    "name": "2x2x2 Cube, Round 1, Group 7",
                    "startTime": "2019-11-23T16:06:00Z",
                  },
                  Object {
                    "activityCode": "222-r1-g9",
                    "childActivities": Array [],
                    "endTime": "2019-11-23T16:30:00Z",
                    "extensions": Array [],
                    "id": 59,
                    "name": "2x2x2 Cube, Round 1, Group 9",
                    "startTime": "2019-11-23T16:18:00Z",
                  },
                ],
                "endTime": "2019-11-23T16:30:00Z",
                "extensions": Array [],
                "id": 2,
                "name": "2x2x2 Cube, Round 1",
                "startTime": "2019-11-23T15:30:00Z",
              },
              Object {
                "activityCode": "555-r1",
                "childActivities": Array [
                  Object {
                    "activityCode": "555-r1-g1",
                    "childActivities": Array [],
                    "endTime": "2019-11-23T17:00:00Z",
                    "extensions": Array [],
                    "id": 71,
                    "name": "5x5x5 Cube, Round 1, Group 1",
                    "startTime": "2019-11-23T16:30:00Z",
                  },
                  Object {
                    "activityCode": "555-r1-g3",
                    "childActivities": Array [],
                    "endTime": "2019-11-23T17:30:00Z",
                    "extensions": Array [],
                    "id": 73,
                    "name": "5x5x5 Cube, Round 1, Group 3",
                    "startTime": "2019-11-23T17:00:00Z",
                  },
                ],
                "endTime": "2019-11-23T17:30:00Z",
                "extensions": Array [],
                "id": 3,
                "name": "5x5x5 Cube, Round 1",
                "startTime": "2019-11-23T16:30:00Z",
              },
            ],
            "color": "#304a96",
            "extensions": Array [],
            "id": 1,
            "name": "Ballroom A - Stage 1",
          },
        ],
        "timezone": "America/New_York",
      },
    ],
  },
  "settings": Object {
    "colorTheme": "orange_200",
    "imageUrl": "",
    "message": "",
  },
  "shortName": "SBU Fall 2019",
}
`;

exports[`Wcif should throw not logged in error 1`] = `
Object {
  "data": null,
  "errors": Array [
    [GraphQLError: Not logged in],
  ],
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;
